### ISSUE TITLE
array aliasing and LLVM performance for elegant tuplesVSarrays

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#11154]
See https://chapel-lang.org/perf/chapcs/llvm/?startdate=2018/07/12&graphs=elegancetuplesvsarrays

Elegant tuplesVSarrays is not performing as well as it could with LLVM. This has to do with detecting array aliasing. Here are some related comments:
 * https://github.com/chapel-lang/chapel/issues/5043#issuecomment-322617090
 * #7762

Issue #7762 notes:

> Array wrapper records do not share the underlying arrays unless they are array views. This could be used to detect easy cases.

So, an initial attempt at solving this problem would be:
 * Chapel compiler can detect when arrays may not alias (e.g. two different array variables that are not array views and neither is a reference)
 * in that event, use noalias and alias.scope metadata to mark arrays that cannot possibly alias

**Comment by bradcray** [comment]
It wasn't clear to me from the notes above whether you were keeping in mind that arrays can also alias via scoping.  For example, in the following A and X alias though neither is an array view:

```chapel
var A: [1..10] real;

proc foo(X: []) {
  A = X;  // A and X are the same thing but neither is an array view
}
```

**Comment by mppf** [comment]
@bradcray - yes that's right, but in that event, they compiler's view is that `X` argument is actually a reference to an array wrapper record. Anyway I'll note it above.

**Comment by mppf** [comment]
First, there is a difference in the "performant" version between --llvm and GCC.
This can be resolved by passing `--no-ieee-float` in which case the LLVM version outperforms the GCC version. This is because the --llvm code generator isn't peephole optimizing vaddpd + vmulpd while the GCC version uses vfmadd132pd (i.e. the GCC version is using fused multiply-adds).

**Comment by mppf** [comment]
Indeed, in the "elegant" version, even with --no-ieee-float, the vfmadd132pd instructions have vmovsd instructions doing loads interspersed.

**Comment by mppf** [comment]
Additional evidence that it's an alias analysis problem is that changing the type of subArray from `real` to `real(32)` resolves the performance problem.

**Comment by mppf** [comment]
PRIM_ARRAY_GET returns a pointer to an array element and might be useful here.
The compiler will need to, after inlining (ideally), decide for each PRIM_ARRAY_GET whether it could possibly alias other PRIM_ARRAY_GETs in the same scope.

**Comment by mppf** [comment]
My branch https://github.com/mppf/chapel/tree/no-alias-sets improves tuplesVSarrays --llvm performance.

on chapcs11, default settings (qthreads), running tuplesVSarrays same as a performance test but with 40 iterations (for added stability in the numbers). Times in seconds:

| config                                                             | elegant                    | performant   |
| ----------------------------                          | ------------------- | ------------ |
|  --fast                                                             | 0.17671                    | 0.209033      |
|  --fast --no-ieee-float                                    | 0.177444                 | 0.209656     |
|  master: --fast --llvm                                     | 1.06911                    | 0.254161       |
|  no-alias-sets: --fast --llvm                           | **0.247053**          | 0.247354     |
|  master: --fast --llvm --no-ieee-float            | 0.678509                | 0.167714     |
|  no-alias-sets: --fast --llvm --no-ieee-float | **0.174716**           | 0.174548     |


What if kernels are in separate procs, rather than inlined?

| config                                                             | elegant                    | performant   |
| ----------------------------                          | ------------------- | ------------ |
|  --fast --no-ieee-float                                    | 0.177348                 | 0.209505     |
|  master: --fast --llvm --no-ieee-float            | 0.687889                | 0.168471     |
|  no-alias-sets: --fast --llvm --no-ieee-float  | **0.175491**          | 0.175374     |

This relies on the interprocedural portion of the pass, but could be also addressed with more aggressive inlining (inline functions with 1 call site, say, even if they are long).

An aside: interestingly, on my home machine, the `--fast --llvm --no-ieee-float` *performant* configuration outperforms the C version. This happens independent of the current effort.

| config                                                             | elegant                    | performant   |
| ----------------------------                          | ------------------- | ------------ |
|  --fast --no-ieee-float                                    | 0.161982                 | 0.161138     |
|  master: --fast --llvm --no-ieee-float            | 0.490201                | **0.105432**     |
|  no-alias-sets: --fast --llvm --no-ieee-float | 0.155945                 | **0.105542**     |

Full command for reproducing:
```
chpl --set subOrder=10 --set order=12000 --set iterations=40 --set correctness=false test/studies/elegance/tuplesVSarrays.chpl  --fast    && ./tuplesVSarrays
```

Experiments were done with commit 1a1aaa5ecd2e55578f5977d2d0d8f9ba3186d529.

**Comment by mppf** [comment]
Closed by #11324.
